src/main/java/htsjdk/tribble/FeatureReader.java

```diff
@@ -27,16 +29,60 @@
  * @param <T> a feature type
  */
 public interface FeatureReader<T extends Feature> extends Closeable {
-    
-    public CloseableTribbleIterator<T> query(final String chr, final int start, final int end) throws IOException;
 
-    public CloseableTribbleIterator<T> iterator() throws IOException;
+    /**
+     * Query the reader for a particular interval corresponding to a contig and a 1-based closed
+     *
+     * @param chr the contig to be queried
+     * @param start the start of the interval (1-based) to be queried
+     * @param end the last base in the interval to be queried
+     * @return an iterator containing the features that at in the interval.
+     * @throws IOException If there's a problem reading or if the reader is not queryable, e.g. if it doesn't have an index.
+     */
+    CloseableTribbleIterator<T> query(final String chr, final int start, final int end) throws IOException;
+
+
+    /**
+     * Query the reader for a particular interval corresponding to a contig and a 1-based closed
+     *
+     * @param locus The locus to be queried
+     * @return an iterator containing the features that at in the interval.
+     * @throws IOException If there's a problem reading or if the reader is not queryable, e.g. if it doesn't have an index.
+     */
+    default CloseableTribbleIterator<T> query(Locatable locus) throws IOException {
+        return query(locus.getContig(), locus.getStart(), locus.getEnd());
+    }
+
+    /**
+     * Provides access to all the features in the reader
+     * @return an iterator to all the features in the reader
+     * @throws IOException If there's a problem reading.
+     */
+    CloseableTribbleIterator<T> iterator() throws IOException;
 
+    /**
+     * Closes the reader
+     * @throws IOException
+     */
     @Override
-    public void close() throws IOException;
+    void close() throws IOException;
 
-    public List<String> getSequenceNames();
+    /**
+     * Provides the list of sequenceNames if known. Otherwise will return an empty list.
+     * @return  the list of sequenceNames if known. Otherwise will return an empty list.
+     */
+    List<String> getSequenceNames();
 
-    public Object getHeader();
+    /**
+     * Provide access to the header of the reader
+     * @return the header of the reader. May be null.
+     */
+    Object getHeader();
 
+    /**
+     * @return true if the reader has an index, which means that it can be queried.
+     */
+    default boolean hasIndex() {
```

Comment:
I think we might want to separate out the concepts of `hasIndex` and `isQueriable`.  In gatk we have a datasource that has no index but supports random access.   I also think `supportsRandomAccess` might be clearer name.  I'm not sure how granular we want to go though, you could potentially imagine a stream with an index that supports fast forward skipping but not rewinding, which wouldn't be easily described here either.

Reply:
hmmm. thats interesting...the ReferenceFile implementation is basically like that...you can move forwards but not backwards in a reference.

Reply:
are you referring to the comment (which ties the two concepts together) or the default implementation (that does the same)?

I think that we should support implementations that exist...and possibly ones that we can easily support...but not implementations that one "could potentially imagine"....

in GATK, is that datasource implemented as a FeatureReader? or on top of one of the standard readers?